front/pages/api/w/[wId]/files/[fileId].ts

```diff
@@ -0,0 +1,217 @@
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message: "Missing fileId query parameter.",
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "GET": {
+      const action: Action = validActions.includes(req.query.action as Action)
+        ? (req.query.action as Action)
+        : "download";
+
+      const filePath = makeStorageFilePathForWorkspaceId(owner, fileId);
+
+      if (action === "view") {
+        const stream = await getPrivateUploadBucket().fetchWithStream(filePath);
+        stream.on("error", () => {
+          return apiError(req, res, {
+            status_code: 400,
+            api_error: {
+              type: "message_not_found",
+              message: "File not found.",
+            },
+          });
+        });
+
+        stream.pipe(res);
+        return;
+      }
+
+      const fileName = await getFileNameFromFileMetadata(filePath);
+
+      // Redirect to a signed URL.
+      const url = await getPrivateUploadBucket().getSignedUrl(filePath, {
+        // Since we redirect, the use is immediate so expiry can be short.
+        expirationDelay: 10 * 1000,
+        promptSaveAs: fileName ?? `dust_${fileId}`,
+      });
+
+      res.redirect(url);
+      return;
+    }
+
+    case "DELETE": {
+      const filePath = makeStorageFilePathForWorkspaceId(owner, fileId);
+
+      await getPrivateUploadBucket().delete(filePath);
+
+      res.status(204).end();
+      return;
+    }
+
+    case "POST": {
+      if (typeof token !== "string") {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: "Invalid token.",
+          },
+        });
+      }
+
+      // Ensure that the file token is valid.
+      const fileTokenPayload = decodeFileToken(token);
+
+      if (!fileTokenPayload || fileTokenPayload.fileId !== fileId) {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: "Invalid file id.",
+          },
+        });
+      }
+
+      try {
+        const form = new IncomingForm();
+        const [, files] = await form.parse(req);
+
+        const maybeFiles = files.file;
+
+        if (!maybeFiles || maybeFiles.length === 0) {
+          return apiError(req, res, {
+            status_code: 400,
+            api_error: {
+              type: "invalid_request_error",
+              message: "No file uploaded",
+            },
+          });
+        }
+
+        const [file] = maybeFiles;
+
+        if (isSupportedImageMimeType(file)) {
+          await resizeAndUploadToFileStorage(owner, fileTokenPayload, file);
+        } else if (isSupportedTextMimeType(file)) {
+          // TODO:(2026-06-28 flav) Move logic to extract text from PDF here.
+          await uploadToFileStorage(owner, fileTokenPayload, file);
```

Comment:
If we're building a new file API we should make sure to support something forward looking. This downloads the entire file in memory before uploading right? 

If not: since we're cutting a new API we want to make sure it supports proper streaming?

Reply:
It creates interesting questions on how we want to handle that: we can't stream and transform at the same time. Which hints that we likely want a file status API and an asyncrhonous processing mechanism run server side to resize images or extract PDFs.